### ISSUE TITLE
Build failure in C11 mode

### ISSUE THREAD

**Issue by ryandesign** [Cyan4973/xxHash#671]
On some systems, xxhash 0.8.1 doesn't compile in C11 mode. This was [reported to MacPorts](https://trac.macports.org/ticket/64113). The error we see is:

```
/usr/bin/clang -Os -arch x86_64 -Wall -Wextra -Wconversion -Wcast-qual -Wcast-align -Wshadow -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wundef -Wpointer-arith -Wformat-security -Wvla -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls -Wstrict-overflow=2  -I/opt/local/include -DXXH_INLINE_ALL cli/xxhsum.c -o xxhsum_inlinedXXH
In file included from xxhash.c:43:
./xxhash.h:2266:5: warning: implicit declaration of function 'static_assert' is invalid in C99 [-Wimplicit-function-declaration]
    XXH_STATIC_ASSERT(sizeof(XXH32_canonical_t) == sizeof(XXH32_hash_t));In file included from xxhash.c:43:
./xxhash.h:2266:5    ^
: warning: ./xxhash.h:1556:32: implicit declaration of function 'static_assert' is invalid in C99 [-Wimplicit-function-declaration]note
: expanded from macro 'XXH_STATIC_ASSERT'
#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE((c),#c)
                               ^
./xxhash.h:1550:54: note: expanded from macro 'XXH_STATIC_ASSERT_WITH_MESSAGE'
#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { static_assert((c),m); } while(0)
                                                     ^
    XXH_STATIC_ASSERT(sizeof(XXH32_canonical_t) == sizeof(XXH32_hash_t));
    ^
./xxhash.h:1556:32: note: expanded from macro 'XXH_STATIC_ASSERT'
#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE((c),#c)
                               ^
./xxhash.h:1550:54: note: expanded from macro 'XXH_STATIC_ASSERT_WITH_MESSAGE'
#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { static_assert((c),m); } while(0)
                                                     ^
In file included from cli/xxhsum.c:41:
In file included from cli/xsum_sanity_check.c:31:
cli/../xxhash.h:2266:5: warning: implicit declaration of function 'static_assert' is invalid in C99 [-Wimplicit-function-declaration]
    XXH_STATIC_ASSERT(sizeof(XXH32_canonical_t) == sizeof(XXH32_hash_t));
    ^
cli/../xxhash.h:1556:32: note: expanded from macro 'XXH_STATIC_ASSERT'
#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE((c),#c)
                               ^
cli/../xxhash.h:1550:54: note: expanded from macro 'XXH_STATIC_ASSERT_WITH_MESSAGE'
#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { static_assert((c),m); } while(0)
                                                     ^
1 warning generated.
1 warning generated.
ar rcs libxxhash.a xxhash.o
/usr/bin/clang -Os -arch x86_64 -Wall -Wextra -Wconversion -Wcast-qual -Wcast-align -Wshadow -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wundef -Wpointer-arith -Wformat-security -Wvla -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls -Wstrict-overflow=2  -I/opt/local/include xxhash.o cli/xxhsum.o cli/xsum_os_specific.o cli/xsum_output.o cli/xsum_sanity_check.o cli/xsum_bench.o -L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64 -o xxhsum
Undefined symbols for architecture x86_64:
  "_static_assert", referenced from:
      _XXH32_canonicalFromHash in xxhash.o
      _XXH64_canonicalFromHash in xxhash.o
Undefined symbols for architecture x86_64:
  "_static_assert", referenced from:
      _XXH32_canonicalFromHash in xxhash-e1ea3c.o
      _XXH3_hashLong_64b_default in xxhash.o
      _XXH3_hashLong_64b_withSecret in xxhash.o
      _XXH3_hashLong_64b_withSeed in xxhash.o
      _XXH3_64bits_withSecretandSeed in xxhash.o
      _XXH3_64bits_reset_withSeed in xxhash.o
      _XXH64_canonicalFromHash in xxhash-e1ea3c.o
      ...
      _XXH3_hashLong_64b_default in xxhash-e1ea3c.o
      _XXH3_hashLong_64b_withSecret in xxhash-e1ea3c.o
      _XXH3_hashLong_64b_withSeed in xxhash-e1ea3c.o
      _XXH3_64bits_withSecretandSeed in xxhash-e1ea3c.o
      _XXH3_64bits_reset_withSeed in xxhash-e1ea3c.o
      ...
ld: symbol(s) not found for architecture x86_64
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [xxhsum] Error 1
make: *** Waiting for unfinished jobs....
make: *** [libxxhash.0.8.1.dylib] Error 1
1 warning generated.
Undefined symbols for architecture x86_64:
  "_static_assert", referenced from:
      _localXXH3_stream_seeded in xxhsum-c9b2d4.o
      _localXXH128_stream_seeded in xxhsum-c9b2d4.o
      _XXH3_hashLong_64b_default in xxhsum-c9b2d4.o
      _XXH3_hashLong_64b_withSeed in xxhsum-c9b2d4.o
      _XXH3_hashLong_64b_withSecret in xxhsum-c9b2d4.o
      _XXH3_hashLong_128b_internal in xxhsum-c9b2d4.o
      _XXH3_hashLong_128b_withSeed in xxhsum-c9b2d4.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [xxhsum_inlinedXXH] Error 1
```

Newer versions of clang default to C11 mode. In MacPorts we worked around the issue by telling xxhash to compile with `-std=c99`, however we then discovered that other software that wants to use xxhash ([such as php](https://trac.macports.org/ticket/64255)) fails in the same way.

It would be great if xxhash could be fixed so that it and whatever uses it can be compiled regardless what language mode the compiler is in.

I assume it is the code that was added in #567 that's causing this problem, but I haven't been able to understand what's wrong. It sure looks to me like the code is intending to use `static_assert` only in C11 or C++11 mode, where it should exist.

**Comment by Cyan4973** [comment]
I believe this issue has been recently fixed in `dev` branch.

As a side quest, I'm wondering why we were not able to generate the same issue during CI tests, since we are supposed to also cover `macos` / `clang` configuration.

**Comment by ryandesign** [comment]
> I believe this issue has been recently fixed

If you mean by 6189ecd3d44a693460f86280ccf49d33cb4b18e1, I agree, that fixes the problem for me. I tested with MacPorts (open source) Clang 9.0.1 on Mac OS X 10.7.5, where previously the build had failed.

> I'm wondering why we were not able to generate the same issue during CI tests, since we are supposed to also cover `macos` / `clang` configuration.

I bet your CI only lets you test with the version of Apple Clang that ships with very recent versions of macOS. In MacPorts, we only saw the problem on Mac OS X 10.6.8 (with MacPorts Clang 9.0.1) and on OS X 10.10.5 (with Apple LLVM version 7.0.2 (clang-700.1.81) from Xcode 7.2.1).
